app/services/order_purchaser.rb

```diff
@@ -0,0 +1,81 @@
+class OrderPurchaser
+  attr_reader :order, :params, :user
+
+  delegate :order_notification_recipient, to: :order
+  delegate :quantities_changed?, to: :order_detail_updater
+
+  def initialize(acting_as:, order:, order_in_past:, params:, user:)
+    @acting_as = acting_as
+    @order = order
+    @order_in_past = order_in_past
+    @params = params
+    @user = user
+  end
+
+  def purchase!
+    order.transaction do
+      return if order_detail_updater.update! && quantities_changed?
```

Comment:
If the quantities have changed, won't that raise an error (which would rollback the transaction)? And the `quantities_changed?` would either not get called or return `false`?